src/operator/tensor/elemwise_unary_op.h

```diff
@@ -31,173 +31,337 @@
-                  const std::vector<TBlob>& inputs,
-                  const std::vector<OpReqType>& req,
-                  const std::vector<TBlob>& outputs) {
-  using namespace mshadow;
-  using namespace mshadow::expr;
-  Stream<xpu> *s = ctx.get_stream<xpu>();
-  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
-    Tensor<xpu, 1, DType> out = outputs[0].FlatTo1D<xpu, DType>(s);
-    ASSIGN_DISPATCH(out, req[0], F<OP>(inputs[0].FlatTo1D<xpu, DType>(s)));
-  });
-}
+  /*! \brief Allocate geometry-related blob data for sparse tensors
+   * \param dest Destination sparse NDArray
+   * \param clone_from sparse NDArray from which to clone storage attributes
+   */
+  static void AllocateGeometry(const NDArray *dest, const NDArray* clone_from = nullptr) {
+    if (clone_from) {
+      const TShape ishape = clone_from->storage_shape();
+      TShape sshape = dest->storage_shape();
+      CHECK(shape_assign(&sshape, ishape));
+      dest->CheckAndAllocData(sshape);
+      CHECK_EQ(dest->storage_type(), clone_from->storage_type());
+      for (size_t i = 0, n = clone_from->aux_shapes().size(); i < n; ++i) {
+        TShape ashape = dest->aux_shape(i);
+        CHECK(shape_assign(&ashape, clone_from->aux_shape(i)));
+        dest->CheckAndAllocAuxData(i, ashape);
+      }
+      DCHECK_EQ(dest->aux_shapes().size(), clone_from->aux_shapes().size());
+    } else {
+      for (size_t i = 0, n = dest->aux_shapes().size(); i < n; ++i) {
+        dest->CheckAndAllocAuxData(i, dest->aux_shape(i));
+      }
+      dest->CheckAndAllocData(dest->storage_shape());
+    }
+  }
 
+  /*! \brief Copy the geometry-related blobs (row sparse indexes, etc.) */
+  template<typename xpu>
+  static inline void CopyGeometryBlobs(mshadow::Stream<xpu> *s,
+                                       const NDArray *dest,
+                                       const OpReqType reqi,
+                                       const NDArray &src) {
+    CHECK_EQ(src.aux_shapes().size(), dest->aux_shapes().size());
+    // My assumption is that the geometry blobs are not large enough to justify an omp loop here,
+    // since the thread synchronization calls for each fork will take longer
+    // than copying a few floats
+    for (size_t i = 0, n = src.aux_shapes().size(); i < n; ++i) {
+      const TBlob src_blob = src.aux_data(i);
+      const TBlob dest_blob = dest->aux_data(i);
+      CopyBlob<xpu>(s, &dest_blob, reqi, src_blob);
+    }
+  }
 
-template<typename xpu>
-void IdentityCompute(const nnvm::NodeAttrs& attrs,
+  /*! \brief Generic copy NDArray */
+  template<typename xpu>
+  static inline void CopyNDArray(mshadow::Stream<xpu> *s,
+                                 const NDArray *dest,
+                                 const OpReqType reqi,
+                                 const NDArray& src) {
+    DCHECK_NE(dest->storage_type(), kDefaultStorage);
+    DCHECK_EQ(dest->storage_type(), src.storage_type());
+    AllocateGeometry(dest, &src);
```

Comment:
It's not just about the write. Let say you read very large csr inputs from a file and train a model by performing symbol.dot(csr_input, weight) with some loss. nnvm will detect that it only want to calculate the grad for `weight` but not for `csr_input`, hence req[csr_input] = kNullOp. In this case we don't even want to allocate memory for grad_csr_input since we don't need it, and csr_input will have huge shape, which can waste lots of memory.